src/st/rwrp/warp.c

```diff
@@ -1,13 +1,351 @@
+            prim->g0 = prim->b0 = prim->r0 += 2;
+        }
+        prim = prim->next;
+        prim->drawMode = DRAW_TRANSP | DRAW_TPAGE | DRAW_TPAGE2;
+        if (prim->r0 < 0xF8) {
+            prim->g0 = prim->b0 = prim->r0 += 2;
+        } else {
+            self->step++;
+        }
+        LOW(prim->r1) = LOW(prim->r0);
+        LOW(prim->r2) = LOW(prim->r0);
+        LOW(prim->r3) = LOW(prim->r0);
+        break;
+    case 5:
+        // Perform the actual warp
+        temp = self->params + 1;
+        for (i = 0; i < LEN(WarpRoomCoords); i++, temp++) {
+            if (temp > LEN(WarpRoomCoords) - 1) {
+                temp = 0;
+            }
+            if (g_CastleFlags[0xD1] & (1 << temp)) {
+                break;
+            }
+        }
+        DestinationWarpRoom = temp;
+        warpCoords = &WarpRoomCoords[temp];
+        moveX = warpCoords->x - g_Tilemap.left;
+        moveY = warpCoords->y - g_Tilemap.top;
+        FntPrint("move_room%x\n", DestinationWarpRoom);
+        FntPrint("for_x:%x y%x\n", warpCoords->x, warpCoords->y);
+        FntPrint("move_x:%x y%x\n", moveX, moveY);
+        PLAYER.posX.i.hi += moveX << 8;
+        PLAYER.posY.i.hi += moveY << 8;
+        self->step = 0x80;
+        D_80097C98 = 2;
+        break;
+    case 6:
+        // start in new room
+        g_Player.D_80072EFC = 0x10;
+        PLAYER.zPriority = 0x94;
+        g_Player.padSim = 0;
+        D_80097408[0] = 0x94;
+        D_8003C8B8 = 0;
+        prim = self->ext.warpRoom.primFade;
+        prim->drawMode = DRAW_HIDE;
+        prim->g0 = prim->b0 = prim->r0 = 0;
+        prim = prim->next;
+        prim->r0 = prim->g0 = prim->b0 = 0xF8;
+        LOW(prim->r1) = LOW(prim->r0);
+        LOW(prim->r2) = LOW(prim->r0);
+        LOW(prim->r3) = LOW(prim->r0);
+        prim->drawMode = DRAW_TRANSP | DRAW_TPAGE | DRAW_TPAGE2;
+        g_api.g_pfn_800EA5AC(0, 0, 0, 0);
+        self->step++;
+        // fallthrough
+    case 7:
+        // Finalize warp by fading in from white
+        g_Player.padSim = 0;
+        g_Player.D_80072EFC = 0x10;
+        D_8003C8B8 = 0;
+        prim = self->ext.warpRoom.primFade;
+        prim = prim->next;
+        temp = prim->r0;
+        temp -= 4;
+        if (temp < 0) {
+            temp = 0;
+            prim->drawMode = DRAW_HIDE;
+            D_8003C8B8 = 1;
+            self->step++;
+        }
+        if (temp < 0x28) {
+            D_80180648 = 0;
+        }
+        prim->g0 = prim->b0 = prim->r0 = temp;
+        LOW(prim->r1) = LOW(prim->r0);
+        LOW(prim->r2) = LOW(prim->r0);
+        LOW(prim->r3) = LOW(prim->r0);
+        break;
+    case 8:
+        // lower alucard up to the room?
+        self->posY.i.hi++;
+        if (collision & 0x4) {
+            g_api.func_8010DFF0(0, 1);
+
+            PLAYER_posY_i_hi += 1;
+            D_80097488.y.i.hi += 1;
+
+            if (g_pads[0].pressed & PAD_UP &&
+                UNDEFINED(GetDistanceToPlayerX)() < 8 &&
```

Comment:
I added the `UNDEFINED` macro because I've seen this pattern several times now. In the past I've written a specific cast or macro for the instance being undefined or cast the function inline, for example: https://github.com/Xeeynamo/sotn-decomp/blob/8cd4eb91bc50dbafda0618f45972fbe41eb489ba/src/st/st0/30030.c#L111

I'm happy to change the name, which may not reflect the intent.

Reply:
Having a simple function prototype will work just fine. I find the `UNDEFINED` confusing.